File: backend/app/reference_text.py
"""Reference-text parsing and validation helpers.

This module mirrors the shipped frontend reference-text semantics without any
framework or persistence dependencies. It keeps token metadata in a registry so
new token kinds can be added without changing the parser.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Generic, Mapping, TypeVar


Context = TypeVar("Context")


@dataclass(frozen=True)
class ReferenceDefinition(Generic[Context]):
    token: str
    kind: str
    domain: str
    fallback: str
    resolve: Callable[[Context], str | int | float | None]


ReferenceRegistry = Mapping[str, ReferenceDefinition[Context]]


def create_reference_registry(
    definitions: list[ReferenceDefinition[Context]] | tuple[ReferenceDefinition[Context], ...],
) -> dict[str, ReferenceDefinition[Context]]:
    registry: dict[str, ReferenceDefinition[Context]] = {}
    for definition in definitions:
        if definition.token in registry:
            raise ValueError(f"Duplicate reference token: {definition.token}")
        registry[definition.token] = definition
    return registry


def _malformed_error(message: str, start: int, end: int) -> dict[str, object]:
    return {
        "kind": "malformed",
        "message": message,
        "start": start,
        "end": end,
    }


def _find_malformed_references(source: str) -> list[dict[str, object]]:
    errors: list[dict[str, object]] = []
    cursor = 0

    while cursor < len(source):
        if source[cursor] == "}":
            errors.append(_malformed_error("Unmatched closing brace", cursor, cursor + 1))
            cursor += 1
            continue

        if source[cursor] != "{":
            cursor += 1
            continue

        close = source.find("}", cursor + 1)
        if close == -1:
            errors.append(_malformed_error("Unmatched opening brace", cursor, len(source)))
            break

        nested_open = source.find("{", cursor + 1)
        if nested_open != -1 and nested_open < close:
            errors.append(_malformed_error("Nested braces are not allowed", cursor, close + 1))
            cursor = close + 1
            continue

        if source[cursor + 1 : close].strip() == "":
            errors.append(_malformed_error("Reference token cannot be empty", cursor, close + 1))
        cursor = close + 1

    return errors


def parse_reference_text(source: str) -> dict[str, object]:
    errors = _find_malformed_references(source)
    if errors:
        return {"valid": False, "errors": errors}

    nodes: list[dict[str, object]] = []
    cursor = 0
    literal_start = 0

    while cursor < len(source):
        if source[cursor] != "{":
            cursor += 1
            continue

        if cursor > literal_start:
            nodes.append(
                {
                    "type": "text",
                    "text": source[literal_start:cursor],
                    "start": literal_start,
                    "end": cursor,
                },
            )

        close = source.find("}", cursor + 1)
        nodes.append(
            {
                "type": "reference",
                "token": source[cursor + 1 : close],
                "source": source[cursor : close + 1],
                "start": cursor,
                "end": close + 1,
            },
        )
        cursor = close + 1
        literal_start = cursor

    if literal_start < len(source):
        nodes.append(
            {
                "type": "text",
                "text": source[literal_start:],
                "start": literal_start,
                "end": len(source),
            },
        )

    return {"valid": True, "document": {"source": source, "nodes": nodes}}


def validate_reference_text(
    source: str,
    registry: ReferenceRegistry[Context],
) -> dict[str, object]:
    parsed = parse_reference_text(source)
    if not parsed["valid"]:
        return parsed

    document = parsed["document"]
    errors: list[dict[str, object]] = []
    for node in document["nodes"]:
        if node["type"] != "reference":
            continue
        if node["token"] in registry:
            continue
        errors.append(
            {
                "kind": "unknown",
                "message": f"Unknown reference token: {node['token']}",
                "token": node["token"],
                "start": node["start"],
                "end": node["end"],
            },
        )

    if errors:
        return {"valid": False, "errors": errors}
    return parsed


def resolve_reference_text(
    document: dict[str, object],
    registry: ReferenceRegistry[Context],
    context: Context,
) -> str:
    parts: list[str] = []
    for node in document["nodes"]:
        if node["type"] == "text":
            parts.append(str(node["text"]))
            continue

        definition = registry.get(str(node["token"]))
        if definition is None:
            parts.append("")
            continue

        value = definition.resolve(context)
        parts.append(definition.fallback if value is None else str(value))

    return "".join(parts)


@dataclass(frozen=True)
class SpellValueReferenceContext:
    spell_attack_bonus: int | float | None = None
    spell_save_dc: int | float | None = None


spell_value_reference_registry = create_reference_registry(
    [
        ReferenceDefinition[SpellValueReferenceContext](
            token="spell_attack_bonus",
            kind="value",
            domain="spell",
            fallback="your spell attack bonus",
            resolve=lambda context: context.spell_attack_bonus,
        ),
        ReferenceDefinition[SpellValueReferenceContext](
            token="spell_save_dc",
            kind="value",
            domain="spell",
            fallback="your spell save DC",
            resolve=lambda context: context.spell_save_dc,
        ),
    ],
)

File: frontend/src/components/referenceText.ts
import { createElement } from "react";
import { DiceText } from "./DiceText";

export interface LiteralReferenceTextNode {
  type: "text";
  text: string;
  start: number;
  end: number;
}

export interface RegisteredReferenceTextNode {
  type: "reference";
  token: string;
  source: string;
  start: number;
  end: number;
}

export type ReferenceTextNode = LiteralReferenceTextNode | RegisteredReferenceTextNode;

export interface ParsedReferenceText {
  source: string;
  nodes: ReferenceTextNode[];
}

export interface ReferenceTextError {
  kind: "malformed" | "unknown";
  message: string;
  start: number;
  end: number;
  token?: string;
}

export type ReferenceTextValidation =
  { valid: true; document: ParsedReferenceText } | { valid: false; errors: ReferenceTextError[] };

export interface ReferenceDefinition<Context> {
  token: string;
  kind: string;
  domain: string;
  fallback: string;
  resolve: (context: Context) => string | number | null | undefined;
}

export type ReferenceRegistry<Context> = ReadonlyMap<string, ReferenceDefinition<Context>>;

export function createReferenceRegistry<Context>(
  definitions: readonly ReferenceDefinition<Context>[],
): ReferenceRegistry<Context> {
  const registry = new Map<string, ReferenceDefinition<Context>>();
  for (const definition of definitions) {
    if (registry.has(definition.token)) {
      throw new Error(`Duplicate reference token: ${definition.token}`);
    }
    registry.set(definition.token, definition);
  }
  return registry;
}

function findMalformedReferences(source: string): ReferenceTextError[] {
  const errors: ReferenceTextError[] = [];
  let cursor = 0;

  while (cursor < source.length) {
    if (source[cursor] === "}") {
      errors.push({
        kind: "malformed",
        message: "Unmatched closing brace",
        start: cursor,
        end: cursor + 1,
      });
      cursor += 1;
      continue;
    }

    if (source[cursor] !== "{") {
      cursor += 1;
      continue;
    }

    const close = source.indexOf("}", cursor + 1);
    if (close === -1) {
      errors.push({
        kind: "malformed",
        message: "Unmatched opening brace",
        start: cursor,
        end: source.length,
      });
      break;
    }

    const nestedOpen = source.indexOf("{", cursor + 1);
    if (nestedOpen !== -1 && nestedOpen < close) {
      errors.push({
        kind: "malformed",
        message: "Nested braces are not allowed",
        start: cursor,
        end: close + 1,
      });
      cursor = close + 1;
      continue;
    }

    if (source.slice(cursor + 1, close).trim().length === 0) {
      errors.push({
        kind: "malformed",
        message: "Reference token cannot be empty",
        start: cursor,
        end: close + 1,
      });
    }
    cursor = close + 1;
  }

  return errors;
}

export function parseReferenceText(source: string): ReferenceTextValidation {
  const errors = findMalformedReferences(source);
  if (errors.length > 0) {
    return { valid: false, errors };
  }

  const nodes: ReferenceTextNode[] = [];
  let cursor = 0;
  let literalStart = 0;

  while (cursor < source.length) {
    if (source[cursor] !== "{") {
      cursor += 1;
      continue;
    }

    if (cursor > literalStart) {
      nodes.push({
        type: "text",
        text: source.slice(literalStart, cursor),
        start: literalStart,
        end: cursor,
      });
    }

    const close = source.indexOf("}", cursor + 1);
    nodes.push({
      type: "reference",
      token: source.slice(cursor + 1, close),
      source: source.slice(cursor, close + 1),
      start: cursor,
      end: close + 1,
    });
    cursor = close + 1;
    literalStart = cursor;
  }

  if (literalStart < source.length) {
    nodes.push({
      type: "text",
      text: source.slice(literalStart),
      start: literalStart,
      end: source.length,
    });
  }

  return { valid: true, document: { source, nodes } };
}

export function validateReferenceText<Context>(
  source: string,
  registry: ReferenceRegistry<Context>,
): ReferenceTextValidation {
  const parsed = parseReferenceText(source);
  if (!parsed.valid) return parsed;

  const errors = parsed.document.nodes
    .filter((node): node is RegisteredReferenceTextNode => node.type === "reference")
    .filter((node) => !registry.has(node.token))
    .map((node): ReferenceTextError => ({
      kind: "unknown",
      message: `Unknown reference token: ${node.token}`,
      token: node.token,
      start: node.start,
      end: node.end,
    }));

  return errors.length > 0 ? { valid: false, errors } : parsed;
}

export function resolveReferenceText<Context>(
  document: ParsedReferenceText,
  registry: ReferenceRegistry<Context>,
  context: Context,
): string {
  return document.nodes
    .map((node) => {
      if (node.type === "text") return node.text;

      const definition = registry.get(node.token);
      if (!definition) return "";
      const value = definition.resolve(context);
      return value === null || value === undefined ? definition.fallback : String(value);
    })
    .join("");
}

export interface ReferenceTextProps<Context> {
  text: string;
  registry: ReferenceRegistry<Context>;
  context: Context;
  invalidFallback?: string;
  role?: string;
}

export const INVALID_REFERENCE_TEXT_FALLBACK = "Reference text unavailable";

export function ReferenceText<Context>({
  text,
  registry,
  context,
  invalidFallback = INVALID_REFERENCE_TEXT_FALLBACK,
  role,
}: ReferenceTextProps<Context>) {
  const result = validateReferenceText(text, registry);
  const resolvedText = result.valid
    ? resolveReferenceText(result.document, registry, context)
    : invalidFallback;

  return createElement(DiceText, { text: resolvedText, role });
}

export interface SpellValueReferenceContext {
  spell_attack_bonus?: number | null;
  spell_save_dc?: number | null;
}

export const spellValueReferenceRegistry = createReferenceRegistry<SpellValueReferenceContext>([
  {
    token: "spell_attack_bonus",
    kind: "value",
    domain: "spell",
    fallback: "your spell attack bonus",
    resolve: (context) => context.spell_attack_bonus,
  },
  {
    token: "spell_save_dc",
    kind: "value",
    domain: "spell",
    fallback: "your spell save DC",
    resolve: (context) => context.spell_save_dc,
  },
]);

export interface WeaponValueReferenceContext {
  weapon_attack_bonus?: number | null;
  weapon_damage_bonus?: number | null;
}

export const weaponValueReferenceRegistry = createReferenceRegistry<WeaponValueReferenceContext>([
  {
    token: "weapon_attack_bonus",
    kind: "value",
    domain: "weapon",
    fallback: "your attack bonus",
    resolve: (context) => context.weapon_attack_bonus,
  },
  {
    token: "weapon_damage_bonus",
    kind: "value",
    domain: "weapon",
    fallback: "your damage bonus",
    resolve: (context) => context.weapon_damage_bonus,
  },
]);
